Upgrade Checkstyle to 10.2 and update checks

yhtMinceraft1010X opened the pull request:
Fixes #124 

There is one additional check that I wanted to implement: [RequireEmptyLineBeforeBlockTagGroup](https://checkstyle.sourceforge.io/config_javadoc.html#RequireEmptyLineBeforeBlockTagGroup). However, this particular line will fail the check:
https://github.com/se-edu/addressbook-level3/blob/865de62540a50cf0690abc36a212eea1e0b4617a/src/main/java/seedu/address/commons/util/CollectionUtil.java#L15-L16

The problem here is that `@see` is treated like any other block tag despite this particular method having no Javadoc description of its own.

yhtMinceraft1010X committed:
Upgrade Checkstyle to 10.2

The latest version of Checkstyle is 10.2. At the same time, the scope
property in JavadocMethod was replaced with accessModifiers in
Checkstyle 8.42.

Let's upgrade Checkstyle and change scope to accessModifiers.

yhtMinceraft1010X committed:
Update Checkstyle checks

Lets update checkstyle rules as follows, and update the code to match.

New rules:
* To match SE-EDU Java Coding Standard (basic):
  1. DefaultComesLast
  2. NoWhitespaceBeforeCaseDefaultColon
  3. AtclauseOrder
  4. JavadocBlockTagLocation
  5. JavadocContentLocationCheck
  6. JavadocMissingLeadingAsterisk
  7. JavadocMissingWhitespaceAfterAsterisk
* To match SE-EDU Java Coding Standard (intermediate):
  1. MethodParamPad
  2. MissingJavadocType
* To match Google's Java Code Style:
  1. JavadocTagContinuationIndentation

Updates to existing rules:
1. Update default throwsIndent value for the Indentation
   from 4 (incorrect) to 8 (correct).

Updates to code to match the above:
1. HelpWindow file: to match JavadocTagContinuationIndentation

damithc commented:
Thanks for this PR @yhtMinceraft1010X 
@se-edu/tech-team-level1 for your review please

LimJunxue reviewed (commented):
Since this is updating the checkstyle file, it might be good to correct the broken link in line 8 as well.
Current link: https://oss-generic.github.io/process/codingstandards/coding-standards-java.html
Correct link: https://oss-generic.github.io/process/codingStandards/CodingStandard-Java.html

yhtMinceraft1010X commented:
> Since this is updating the checkstyle file, it might be good to correct the broken link in line 8 as well. Current link: https://oss-generic.github.io/process/codingstandards/coding-standards-java.html Correct link: https://oss-generic.github.io/process/codingStandards/CodingStandard-Java.html

Is there any difference between https://oss-generic.github.io/process/codingStandards/CodingStandard-Java.html and the [SE-EDU guide](https://se-education.org/guides/conventions/java/intermediate.html)? I noticed the former link was last updated in 2018 and I think the SE-EDU version is more suitable.

damithc commented:
> Is there any difference between https://oss-generic.github.io/process/codingStandards/CodingStandard-Java.html and the [SE-EDU guide](https://se-education.org/guides/conventions/java/intermediate.html)? I noticed the former link was last updated in 2018 and I think the SE-EDU version is more suitable.

oss-generic is now defunct. Use se-education.org/guides instead.

yhtMinceraft1010X committed:
Update coding standard link and description

The oss-generic link for the coding standard is obsolete. At the same
time, the description vaguely refers to a module.

Let's change the coding standard link to the one used by SE-EDU and
remove the reference to the module.

yhtMinceraft1010X committed:
Update check comments

The JavadocType and JavadocMethod checks are meant to check that the
Javadoc comments follow the appropriate format. On the other hand, the
MissingJavadocMethodCheck checks whether the method has missing Javadoc.

Let's update the comments to describe these checks.

damithc commented:
Getting there 👍 

Minor:  the current commit message for v3 2/4:

```
There are some checks which enforce the Java basic and intermediate
coding standard as well as the Google Java Style Guide. As an example,
the MethodParamPad check ensures that a method or constructor name is
attached to its open parenthesis. The JavadocTagContinuationIndentation
check is for the Google style guide rule that continuation lines for
block tags are indented 4 or more whitespaces relative to the @.

The default throwsIndent value for the Indentation check is 4
whitespaces although the correct value should be 8.

Let's update the throwsIndent value, add the new checks and fix the
HelpWindow file for the JavadocTagContinuationIndentation check.
```

@yhtMinceraft1010X  The justification of the rules can be more systematic. Perhaps revise the first paragraph to justify addition of each rule e.g., refer to the corresponding rule in the SE-EDU coding standard? Alternatively, list which were added on account of the basic rules, which for intermediate rules, which for Google style guide.

You can post the proposed commit message here for comments before updating the actual commit message.

yhtMinceraft1010X commented:
> You can post the proposed commit message here for comments before updating the actual commit message.

Here is my proposed commit message for v3 2/4:
```
There are some checks which enforce the Java basic and intermediate
coding standard as well as the Google Java Style Guide. 

The DefaultComesLast check ensures that default is the last statement 
for switch. The NoWhitespaceBeforeCaseDefaultColon check ensures 
that there is no whitespace between case and the colon or default and 
the colon. Both reference the example code for the basic rule on switch 
statement form.

The MethodParamPad check ensures that a method or constructor name 
is attached to its open parenthesis under the intermediate rule for line 
breaks.

The JavadocTagContinuationIndentation check is for the Google style 
guide rule 7.1.3: continuation lines for block tags are indented 4 or 
more whitespaces relative to the @.

All other newly added Javadoc checks are for the basic rule on the form 
of Javadoc comments.

The default throwsIndent value for the Indentation check is 4
whitespaces although the correct value should be 8.

Let's update the throwsIndent value, add the new checks and fix the
HelpWindow file for the JavadocTagContinuationIndentation check.
```

damithc commented:
We can shorten like this (no need to use paragraph form if there is a better format):
```
Lets update checkstyle rules as follows, and update the code to match.

New rules:
* To match SE-EDU Java Coding Standard (basic):
  1. DefaultComesLast 
  2. NoWhitespaceBeforeCaseDefaultColon
  3. ...
* To match SE-EDU Java Coding Standard (intermediate):
  1. ...
* To match Google's Java Code Style:
  1. JavadocTagContinuationIndentation  

Updates to existing rules:
1. Update default throwsIndent value for the Indentation
   from 4 (incorrect) to 8 (correct).

Updates to code to match the above:
1. HelpWindow file: to match JavadocTagContinuationIndentation
```

damithc commented:
To add, the explanation of each rule can be in the file itself, rather than the commit message.

damithc commented:
@se-edu/tech-team-level1 for your review please ...

Sorry @yhtMinceraft1010X , I missed this earlier.

damithc commented:
Side note: As the four commits seem to be independent changes, I guess we can use 'rebase and merge' (i.e., no merge commit)?

fans2619 reviewed (approved):
LGTM.

vimuthm commented:
> Side note: As the four commits seem to be independent changes, I guess we can use 'rebase and merge' (i.e., no merge commit)?

Yes, a rebase makes sense
